### PR TITLE
Remove for loop trick for inlining functions w/o multiple returns + optims

### PR DESCRIPTION
**Pull request by SteveBronder**
#### Summary

This PR removes for loops surrounding inlined functions if the inlined function does not have more than one return statement while also allow C++ to perform move semantics from the inlined functions inner return to it's outer return object.

After @WardBrian posted #1096 I thought it was odd that inlined functions are always wrapped in a for loop with breaks at the  bottom. So this PR looks at a function as we are inlining it and if the function only has one return statement then the inlined function's code is placed into a `Block` (curly braces) instead of a for loop. For instance the following code is wrapped inside of a for loop that just runs once.

```cpp
        double inline_sym1__;
        for (int inline_sym3__ = 1; inline_sym3__ <= 1; ++inline_sym3__) {
          inline_sym1__ = (blah(1.0, pstream__) + 0.5);   
          break;
        }
```

But we can see that there are not multiple returns here and so instead of having a for loop we can just have a block

```cpp
        double inline_sym1__;
        {
          inline_sym1__ = (blah(1.0, pstream__) + 0.5);   
        }
```

After cleaning that up, I noticed that the inlining code did a pattern where an  variable is declared outside of the block scope, than another variable is declared at the inlined functions scope, and at the end of the inlined functions scope the inner variable is assigned to the outer variable. The below is pseudocode showing what I mean here.

```cpp
Eigen::Matrix<local_scalar_t__, -1, -1> A_outer;
{
  Eigen::Matrix<local_scalar_t__, -1, -1> A_inner;
  // Do a bunch of stuff with A_inner ...
  assign(A_outer, A_inner);
}
```

This is the exact kind of place where we want to do move semantics, where at the end of the scope instead of copying `A_inner` to `A_outer` we can just hand `A_outer` the memory of `A_inner`. This is safe and good because `A_inner` will be deleted at the end of the inlined functions block scope. To let the C++ know an expression can be moved I made an internal `fun_kind` called `FnMove` where if an expression is wrapped in a `FnMove` the C++ code generator can then call `std::move(expr)` such as the below.

```cpp
Eigen::Matrix<local_scalar_t__, -1, -1> A_outer;
{
  Eigen::Matrix<local_scalar_t__, -1, -1> A_inner;
  // Do a bunch of stuff with A_inner ...
  assign(A_outer, std::move(A_inner));
}
```

While I was reading this PR I was having a hard time following which `inline_sym{number}__` was doing what and where. So now when we generate code for inline functions the local inlined variables have the pattern `inline_{function}_{orig_name}_sym{number}__`. The outer return mentioned above now has the name `inline_{func_name}_return_sym{number}__`. The iterator for the for loop trick when we have multiple returns now has the pattern `inline_{func_name}_{iterator}_sym{number}__`.

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Release notes

Replace this text with a short note on what will change if this pull request is merged. This will be included in the release notes.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/analysis_and_optimization/Optimize.ml
Comment (WardBrian) [inline src/analysis_and_optimization/Optimize.ml:127]: I think this function could just be a `Stmt.Fixed.fold`?

File: src/analysis_and_optimization/Optimize.ml
Comment (WardBrian) [inline src/analysis_and_optimization/Optimize.ml:146]: I'd prefer this be a guard clause in the match rather than this big if, so you have two branches:
```ocaml
| Some name, Some e when num_returns > 1 -> ... contents of if
| Some name, Some e -> ... contents of else
```

Similar for the other cases where the body of the match is just an if statement

**Review by WardBrian — COMMENTED** [review]
A few comments questions. The test output on this looks a lot cleaner than the existing stuff!

File: src/middle/Internal_fun.ml
Comment (SteveBronder) [reply src/middle/Internal_fun.ml:28]: Yeah it's very C++ specific in that most other languages do not use RAII (Resource Allocation Is Initialization) with deterministic destructors for classes. 

One way I've been thinking of is kind of a big change. Essentially I'd like the optims to be able to tell the code gen extra information they can choose to use. So for instance below we have an additional `Info` type for `Meta`'s record that for `Var` stores whether the var is "movable" or "lower_triangular" (I'm not saying that is how we should do triangularity fyi it's just an example)

```ocaml
module Typed = struct
  module Meta = struct
    module Info = struct 
      type t =
      | Var of { movable: bool; lower_triangular : bool }
      | Lit 
      | FunApp 
      | TernaryIf 
      | EAnd 
      | EOr 
      | Indexed 
      | PromotionMeta
      [@@deriving sexp, hash, map, fold, compare]
    end
    type t =
      { type_: UnsizedType.t
      ; loc: (Location_span.t[@sexp.opaque] [@compare.ignore])
      ; adlevel: UnsizedType.autodifftype
      ; info : Info.t }
    [@@deriving compare, create, sexp, hash]
...
```

This change would also mean that where we currently have strings for function argument inputs we would need to make them `Var` types. . So the optimizer can go through and say, "This is the last time we use object X so after this call we are free to move the object so mark `movable` as true". Or when doing the transformed MIR we can look for which objects are type `LowerTriangular` and then for code gen mark their associated `Var` types in functions with `lower_triangular = true` so we can print them as `{my_obj}.triangularView<Eigen::Lower>()`. 

The benefit of this is that code gens could always ignore this extra information and still get a compilable program.

But imo I think this is a bigger project we can do at some point. The `FnMove` is pretty similar to internal compiler things like `MakeArray` or even the `Promoter` expression in that it's kind of C++ specific (an R code gen would just done the promotion!). So imo I think it's okie dokie. At worst we can just deprecate it in the future.

File: src/analysis_and_optimization/Optimize.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Optimize.ml:127]: Sorry I think I'm just unfamiliar, what would that look like?

File: src/analysis_and_optimization/Optimize.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Optimize.ml:146]: Much cleaner!

File: src/analysis_and_optimization/Optimize.ml
Comment (WardBrian) [reply src/analysis_and_optimization/Optimize.ml:127]: Folding functions will walk over a data structure and call a two argument function on each node. In some languages this is called `reduce` instead of `fold`. The first argument will be the "accumulator", the second will be the node. Since our Stmt actually has two types inside (the expression type and then the statement type itself) you give two functions, one for each.

The result is something like  

```ocaml
let count_returns = 
Stmt.Fixed.fold  (fun acc _ -> acc) 
(fun acc -> function Stmt.Fixed.{pattern=Return _; _} -> acc + 1 | _ -> acc) 
0
```
(I don't have the source in front of me to make sure that compiles, but it should be close)

It basically avoids the need for defining lots of functions that look similar to what you've defined that mostly just walk over all the sub nodes. There are a bunch of usages of it around the optimize code like here:

https://github.com/stan-dev/stanc3/blob/d2358be75753fc13825c5fd45e03c99e67d2011c/src/analysis_and_optimization/Factor_graph.ml#L37-L44

**Comment by SteveBronder** [comment]
Also fyi before we merge this I think I should turn it on for all of the tests that get compiled and run it once through Jenkins to help ensure there are no weird edge cases we are missing

**Comment by WardBrian** [comment]
Do you want this in before the RC? If it can wait we have tons of time to do that

**Comment by SteveBronder** [comment]
It can absolutely wait. Yeah thinking it over that would be good to do. Though I may need some help as that sort of schema is a little over what I'm used to doing

File: test/integration/good/compiler-optimizations/cppO1.expected
Comment (SteveBronder) [inline test/integration/good/compiler-optimizations/cppO1.expected:12441]: So, I think the DCE scheme is not actually deleting all non-used objects? Else this would be deleted right?

**Comment by WardBrian** [comment]
Is there a reason to update Core? We can, but I’ll need to do another windows port and our CI will need updates

**Comment by SteveBronder** [comment]
No we don't as long as you can tell me how to downgrade the stuff on my laptop after running `opam upgrade` lol. That's the only reason these are updated

**Comment by WardBrian** [comment]
You should be able to do `opam pin core_kernel=0.14.2`

**Comment by SteveBronder** [comment]
`opam pin add core_kernel v0.14.2` did the trick! I think we should add that and `opam pin add fmt 0.8.8` to the setup scripts

**Comment by WardBrian** [comment]
We could, though I think in general there is no reason to run `opam upgrade` in the switch you use for Stan. Effectively everything is pinned

File: src/analysis_and_optimization/Optimize.ml
Comment (WardBrian) [inline src/analysis_and_optimization/Optimize.ml]: Rename to something other than blah

File: src/analysis_and_optimization/Optimize.ml
Comment (WardBrian) [inline src/analysis_and_optimization/Optimize.ml:281]: This needs a lot of explanation in the code I think. Both what is doing, why it is needed, why size 0 is safe, etc

Comment (WardBrian) [reply src/analysis_and_optimization/Optimize.ml:281]: Possibly put it in SizedType.ml (or unsizedtype) with a name like `unsafe_unsized_to_sized`

**Review by WardBrian — COMMENTED** [review]
A few style/organization comments. I think that the logic here is sound in general, but I'd appreciate some clarifying comments in the code, especially around the sized 0 sizedtypes created.

File: src/analysis_and_optimization/Optimize.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Optimize.ml]: I moved this over to `Mir_utils` as I think this is a thing we are only going to use in the optimizations. Also added a bunch of docs describing the behavior since it is an odd function

**Comment by SteveBronder** [comment]
Ty for looking this over! lmk if you have more Qs or suggestions

**Review by WardBrian — APPROVED** [review]
This looks good to me. I propose squashing and merging rather than just a normal merge, but I leave that up to you @SteveBronder

**Comment by SteveBronder** [comment]
Sounds good! I'll merge on Monday

**Pull request merged**
